### PR TITLE
fix: search should not panic on empty table

### PR DESCRIPTION
**Pull request by chebbyChefNEQ**
This PR add a test for KNN search on an empty table. Our code previously returns error because `KNNFlat` returned an empty record batch without `_distance` column. This PR makes `KNNFlat` return a schema with `_distance` column.

This PR also adds a test for KNN search on an empty table that was previously non-empty. This also seems to fail, but is not fixed in this PR. I created a new issue [here](https://github.com/lancedb/lance/issues/1288)

This PR sneaks in `location: Location` field to `crate::Error::Schema`. I needed this for debugging.

### CODE REVIEW

File: rust/lance/src/index/vector/flat.rs [hunk 1]
```diff
@@ -72,7 +75,8 @@ pub async fn flat_search(stream: impl RecordBatchStream, query: &Query) -> Resul
         .await?;
 
     if batches.is_empty() {
-        return Ok(RecordBatch::new_empty(input_schema));
+        let schema_with_distance = input_schema.try_with_column(distance_field())?;
+        return Ok(RecordBatch::new_empty(schema_with_distance.into()));
```
Comment (westonpace) [inline rust/lance/src/index/vector/flat.rs:79]: Does you know if we include the row id on empty tables if the user asks for the row id?

File: rust/lance/src/dataset.rs [hunk 2]
```diff
@@ -2421,4 +2421,93 @@ mod tests {
...
+                &Float32Array::from_iter_values((0..128).map(|_| 0.1)),
+                1,
+            )
+            .unwrap()
+            .try_into_stream()
+            .await
+            .unwrap();
+
+        while let Some(batch) = stream.next().await {
+            std::mem::drop(batch.unwrap())
```
Comment (westonpace) [inline rust/lance/src/dataset.rs:2462]: Should you test the schema here to ensure it has the distance column?

**Review by westonpace — APPROVED** [review]
Good fix.  I appreciate adding location to the errors.

Not really important for this PR but From snafu's docs it looks like location can be implicitly added in some cases but I don't know if there is some reason that wouldn't work for us (it mentions needing the rust 1.46 feature flag and potential issues with async where it reports the location the future was awaited).

However, the overhead of adding `location!()` to every error isn't really that significant so not important.

File: rust/lance/src/index/vector/flat.rs (hunk 1 above)
Comment (chebbyChefNEQ) [reply rust/lance/src/index/vector/flat.rs:79]: for this one, I think row id will either come from the `Take` node. We won't have to handle appending row id here. I'll check if `Take` handles this correctly.
